### PR TITLE
Fix `PhimoeIntegrationTest`

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

This test just get the pytest process being killed due to CPU OOM.

```
tests/models/phimoe/test_modeling_phimoe.py::PhimoeIntegrationTest::test_model_phimoe_instruct_logits 
-------------------------------- live log call ---------------------------------
WARNING  transformers.modeling_rope_utils:modeling_rope_utils.py:418 Unrecognized keys in `rope_scaling` for 'rope_type'='longrope': {'short_mscale', 'long_mscale'}
WARNING  transformers.modeling_rope_utils:logging.py:328 This model has set a `original_max_position_embeddings` field, to be used together with `max_position_embeddings` to determine a scaling factor. Please set the `factor` field of `rope_scaling`with this ratio instead -- we recommend the use of this field over `original_max_position_embeddings`, as it is compatible with most model architectures.
Killed
```

Change to

> dtype="auto", device_map="auto"

helps, but `test_phimoe_instruct_with_static_cache` still get GPU OOM even with super short inputs.

So I also change that test to use a dummy model. There is nothing we can do.

### CODE REVIEW

File: tests/models/phimoe/test_modeling_phimoe.py [hunk 1]
```diff
@@ -166,17 +184,22 @@ def test_phimoe_instruct_generation(self):
...
         self.assertListEqual(output_text, EXPECTED_OUTPUT)
 
     def test_phimoe_instruct_with_static_cache(self):
-        model = PhimoeForCausalLM.from_pretrained("microsoft/Phi-3.5-MoE-instruct")
+        model = self.get_model()
+        # Can't run with the real checkpoint, even if offloaded. Let's just use a tiny dummy one
+        config = copy.deepcopy(model.config)
+        config.num_hidden_layers = 2
+        torch.manual_seed(42)
+        model = type(model)(config)
```
Comment (ydshieh) [inline tests/models/phimoe/test_modeling_phimoe.py]: dummy model

File: tests/models/phimoe/test_modeling_phimoe.py [hunk 2]
```diff
@@ -130,31 +132,47 @@ def test_model_rope_scaling_from_config(self, scaling_type):
...
+    def test_model_phimoe_instruct_logits(self):
+        input_ids = {"input_ids": torch.tensor([[1212, 318, 281, 1672]], dtype=torch.long, device=torch_device)}
 
-        model = PhimoeForCausalLM.from_pretrained("microsoft/Phi-3.5-MoE-instruct").to(torch_device)
+        model = self.get_model()
         model.eval()
 
-        output = model(**input_ids).logits
+        with torch.no_grad():
+            output = model(**input_ids).logits
```
Comment (vasqu) [inline tests/models/phimoe/test_modeling_phimoe.py:163]: Just fmi, we shouldn't need this in eval mode, no? But doesnt hurt, more of a question around my understanding

File: tests/models/phimoe/test_modeling_phimoe.py [hunk 3]
```diff
@@ -130,31 +132,47 @@ def test_model_rope_scaling_from_config(self, scaling_type):
...
-          1.5781,  0.3535,  0.0874,  0.6836, -0.0518, -1.2969]]).to(torch_device)  # fmt: skip
+        EXPECTED_OUTPUT = torch.tensor(
+            [
+                    [-3.4844, -2.4531, -1.1719, 0.6055, -0.4922, -0.1001, 0.8086, -0.2422, 0.3477, -1.0078],
+                    [-0.9766, 0.1631, -0.5508, 2.3594, 0.7031, 3.1719, 0.4141, 0.2305, 0.6055, -2.1250],
+            ]
+        ).to(device=torch_device, dtype=output.dtype)  # fmt: skip
 
-        torch.testing.assert_close(EXPECTED_OUTPUT, output[0, :2, :30], rtol=1e-4, atol=1e-4)
+        torch.testing.assert_close(output[0, :2, :10], EXPECTED_OUTPUT, rtol=1e-4, atol=1e-4)
```
Comment (vasqu) [inline tests/models/phimoe/test_modeling_phimoe.py:172]: Hmm, not a fan of reducing the amount here in general - is it just to have faster tests?

Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py:172]: Same for the instruct test

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 1 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: Fair but let's call the `Phimoe` class directly. Idk but `type(model)` is ugly / bad practice imo :sweat_smile:

File: tests/models/phimoe/test_modeling_phimoe.py [hunk 4]
```diff
@@ -188,12 +211,13 @@ def test_phimoe_instruct_with_static_cache(self):
         ]
         inputs = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt")
 
-        response_tokens = PhimoeMiniWithStaticCache.generate(model, inputs, 64)
+        response_tokens = PhimoeMiniWithStaticCache.generate(model, inputs, max_seq_len=10)
```
Comment (vasqu) [inline tests/models/phimoe/test_modeling_phimoe.py]: Isn't it just one (or a few) tokens generated then. Can we increase the output by a bit at least?

**Review by vasqu — COMMENTED** [review]
Overall fine with the changes especially like the proper setups and cleanups :pray: 

My gripe is mostly that we shouldnt reduce the amount of produced tokens by that much

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 2 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py:163]: Need `torch.no_grad` otherwise there will be gradient , even if we have `eval`, and GPU OOM.

`model.eval()` is mainly for dealing with dropout and something like batch norm layers (compute the statistic or use the computed one )

For this test itself, probably doesn't matter much (for `eval` or not), but it's just more a practice for inference.

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 3 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py:172]: this is not generation , just a forward, and I just set the number of place to compare against.

It doesn't matter much of compare 30 places or 10 places in the tensor (the last dimension)

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 2 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py:163]: Gotcha thanks for the explanation :D

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py]: we are using dummy random model (with fixed seed), there is no point to have more tokens and compare the results here

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 3 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py:172]: For instruct, because we are offload to cpu/disk, the process is really slow.

For this model, might still be able to keep 30 , it runs in 0:02:47.

(there are some other model like glm4v_moe where it's way too slow, 3 tokens takes 16 minutes)

File: tests/models/phimoe/test_modeling_phimoe.py [hunk 5]
```diff
@@ -166,17 +184,29 @@ def test_phimoe_instruct_generation(self):
...
+        config = copy.deepcopy(model.config)
+        config.num_hidden_layers = 2
+        # make `head_dim = 128`
+        config.hidden_size = 512
+        config.num_attention_heads = 4
+        config.num_key_value_heads = 1
+        config.intermediate_size = 512
+        config.max_position_embeddinqgs = 64
+        config.num_local_experts = 4
+        torch.manual_seed(42)
```
Comment (ydshieh) [inline tests/models/phimoe/test_modeling_phimoe.py:207]: have to reduce the dim otherwise still get GPU OOM with 2 layers

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 4 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: I just want to make sure that we can at least generate a few tokens (e.g. 5) even if it is a dummy model. The reason is more so that the static cache goes through the prefill and decoding for a few steps and doesnt break - not the output itself.

Also, can we use `max_new_tokens` instead then? With `max_seq_len`, I need to know the token count myself to calculate the output amount myself

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 3 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py:172]: Gotcha fair point

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 5 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py:207]: Let's add a TODO as well, this should be rather temporary and not a long term solution. There seems to be something fishy going on imo

**Review by vasqu — APPROVED** [review]
Preapproving since the last points are nothing big, thx for bearing with me :p

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py]: I can change to 30 here. It's already 10 which is larger than 5.

Regarding

> max_seq_len

    class PhimoeMiniWithStaticCache(torch.nn.Module):

        @staticmethod
        def generate(model: PhimoeForCausalLM, prompt_tokens: torch.LongTensor, max_seq_len: int) -> list[int]:
            model = PhimoeMiniWithStaticCache(model, 1, max_seq_len + prompt_tokens.shape[-1])


It behaves as max_new_tokens.

I have no idea why we define `PhimoeMiniWithStaticCache` in the test file however, but I don't want to spend any more time on this PR.

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 5 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py:207]: There is nothing fishy  here. It's just the model is too large to fit into A10 (during forward)

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 4 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: Oh damn, I didn't notice it was a separate class... Could it be that the gradients are also calculated and thus the memory usage shoots up? And thx for the clearup, that was a misunderstanding of mine with the new class.

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 5 above)
Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py:207]: Can you check re https://github.com/huggingface/transformers/pull/41007#discussion_r2363556135

Other than that it should be good then :+1: Just a hunch, could be wrong

File: tests/models/phimoe/test_modeling_phimoe.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py]: change it to 30 but the generation just stop early.

Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: > Could it be that the gradients are also calculated and thus the memory usage shoots up?

Can you check this? With torch no grad or similar?

Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: > change it to 30 but the generation just stop early.

Thats on me I didnt notice it was a separate class :cry: it was fine before as well, I thought it just generated like 1 token - Im getting tired argh

Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py]: oh, good point on graident! let me check, you are so goooood!

Comment (ydshieh) [reply tests/models/phimoe/test_modeling_phimoe.py]: with it, no more GPU OOM, but yeah, only generate like one token. I will change back to using the real model however, but won't spend time on this custom class defined in the test file.

Comment (vasqu) [reply tests/models/phimoe/test_modeling_phimoe.py]: Yea no worries no need to invest too deeply there. But that explains the overly high memory usage :D
